ENIGMAsystem/SHELL/Universal_System/Serialization/type_traits.h

```diff
@@ -0,0 +1,268 @@
+/** Copyright (C) 2022 Dhruv Chawla
+*** Copyright (C) 2023 Fares Atef
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+/**
+  @file type_traits.h
+  @brief This file contains type traits which are used with the (de)serialization functions.
+*/
+
+#ifndef ENIGMA_TYPE_TRAITS_H
+#define ENIGMA_TYPE_TRAITS_H
+
+#include <complex>
+#include <queue>
+#include <set>
+#include <stack>
+#include "../var4.h"
+
+template <typename T>
+struct is_lua_table : std::false_type {};
+
+template <typename U>
+struct is_lua_table<lua_table<U>> : std::true_type {
+  using inner_type = U;
+};
+
+template <typename T>
+struct lua_inner_type;
+
+template <typename T>
+struct lua_inner_type<lua_table<T>> {
+  using type = T;
+};
+
+template <typename T>
+constexpr static inline bool is_lua_table_v = is_lua_table<T>::value;
+
+template <typename T>
+struct is_std_vector : std::false_type {};
+
+template <typename T, typename Alloc>
+struct is_std_vector<std::vector<T, Alloc>> : std::true_type {};
+
+template <typename T>
+constexpr static inline bool is_std_vector_v = is_std_vector<T>::value;
+
+template <typename T>
+struct is_std_set : std::false_type {};
+
+template <typename Key, typename Compare, typename Allocator>
+struct is_std_set<std::set<Key, Compare, Allocator>> : std::true_type {};
+
+template <typename T>
+constexpr bool is_std_set_v = is_std_set<T>::value;
+
+template <typename T>
+struct is_std_queue : std::false_type {};
+
+template <typename T, typename Container>
+struct is_std_queue<std::queue<T, Container>> : std::true_type {};
+
+template <typename T>
+constexpr bool is_std_queue_v = is_std_queue<T>::value;
+
+template <typename T>
```

Comment:
tbh I would use macros for this. It's fallen out of style, though.

Have you tried a generic `is_container<T, container>`? I believe you could do this with `std::is_same` (and FYI you can instantiate dependent templates if you say `template` before the class you want the parser to treat as a template).

Reply:
okay, nice idea.